in_toto/runlib_test.go

```diff
@@ -27,6 +28,117 @@ func TestRecordArtifact(t *testing.T) {
+// We will only calculate the hash for for.tar.gz
+// The symlink will not be added to the list right now, nor will we calculate a checksum for it.
+func TestSymlinkToFile(t *testing.T) {
+	if err := os.Symlink("foo.tar.gz", "foo.tar.gz.sym"); err != nil {
+		t.Errorf("Could not create a symlink: %s", err)
+	}
+
+	expected := map[string]interface{}{
+		"foo.tar.gz": map[string]interface{}{
+			"sha256": "52947cb78b91ad01fe81cd6aef42d1f6817e92b9e6936c1e5aabb7c98514f355",
+		},
+	}
+	result, err := RecordArtifacts([]string{"foo.tar.gz.sym"}, 0)
+	if !reflect.DeepEqual(result, expected) {
+		t.Errorf("RecordArtifacts returned '(%s, %s)', expected '(%s, nil)'",
+			result, err, expected)
+	}
+
+	if err := os.Remove("foo.tar.gz.sym"); err != nil {
+		t.Errorf("Could not remove foo.tar.gz.sym: %s", err)
+	}
+}
+
+// TestSymlinkToFolder checks if we are successfully following symlinks to folders
+func TestSymlinkToFolder(t *testing.T) {
+	if err := os.MkdirAll("symTest/symTest2", 0700); err != nil {
+		t.Errorf("Could not create tmpdir: %s", err)
+	}
+
+	if err := os.Symlink("symTest/symTest2", "symTmpfile.sym"); err != nil {
+		t.Errorf("Could not create a symlink: %s", err)
+	}
+
+	// create a filepath from slash, because otherwise
+	// our tests are going to fail, because the path matching will
+	// not work correctly on Windows
+	p := filepath.FromSlash("symTest/symTest2/symTmpfile")
+
+	if err := ioutil.WriteFile(p, []byte("abc"), 0400); err != nil {
+		t.Errorf("Could not write symTmpfile: %s", err)
+	}
+
+	result, err := RecordArtifacts([]string{"symTmpfile.sym"}, 0)
+	if err != nil {
+		t.Error(err)
+	}
+
+	expected := map[string]interface{}{
+		p: map[string]interface{}{
+			"sha256": "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad",
+		},
+	}
+
+	if !reflect.DeepEqual(result, expected) {
+		t.Errorf("RecordArtifacts returned '(%s, %s)', expected '(%s, nil)'",
+			result, err, expected)
+	}
+
+	// make sure to clean up everything
+	if err := os.Remove("symTest/symTest2/symTmpfile"); err != nil {
+		t.Errorf("Could not remove path symTest/symTest2/symTmpfile: %s", err)
+	}
+
+	if err := os.Remove("symTmpfile.sym"); err != nil {
+		t.Errorf("Could not remove path symTest/symTest2/symTmpfile.sym: %s", err)
+	}
+
+	if err := os.Remove("symTest/symTest2"); err != nil {
+		t.Errorf("Could not remove path symTest/symTest2: %s", err)
+	}
+
+	if err := os.Remove("symTest/"); err != nil {
+		t.Errorf("Could not remove path symTest: %s", err)
+	}
+}
+
+// This test provokes a symlink cycle
+func TestSymlinkCycle(t *testing.T) {
+	if err := os.Mkdir("symlinkCycle/", 0700); err != nil {
+		t.Errorf("Could not create tmpdir: %s", err)
+	}
+
+	// we need to get the current working directory here, otherwise
+	// os.Symlink() will create a wrong symlink
+	dir, err := os.Getwd()
+	if err != nil {
+		t.Error(err)
+	}
+	// create a cycle ./symlinkCycle/symCycle.sym -> ./symlinkCycle/
```

Comment:
Would you mind adding a test for an indirect circle? E.g.
/bar/foo.sym --> /foo/ 
/foo/bar.sym --> /bar/

Reply:
This makes sense

Reply:
I've added another test for your indirect cycle.